### DISCUSSION TITLE
Takeaways from  Immutable Collections article

### DISCUSSION THREAD

**Discussion by Aryan-sharma11** [Aryan-sharma11/AddSevenPS#19] [General]
Article link - https://docs.microsoft.com/en-us/archive/msdn-magazine/2017/march/net-framework-immutable-collections

Some of the takeaways from articles on Immutable collections are 

- Immutability is one of the principles of a successful programming paradigm

- F# has all instances immutable by default unless explicitly specified

- These  are collections of instances that preserve their structure all the time and disallow element level assignments

- String is an immutable structure as the new copy is made every time a change is made, whereas a string builder is immutable. 

-  Using string builder for immutable purposes will require locks

-  Pushing elements in an immutable stack creates two stacks 1 with the new element and the other without it

- String and Int32 are immutable whereas String builder is mutable

- Every object during initialization is mutable, once a reference to the object is published, it becomes immutable for the rest of its lifetime

- Immutable stacks are implemented using Linked lists,  immutable lists are implemented using balanced binary trees, only immutable arrays are implemented using arrays

**Comment by abhi18av** [comment]
Please add the resources you've used to arrive to this conclusion.

**Comment by Aryan-sharma11** [comment]
Working on Immutable lists using balanced binary trees

- To add the first item to the tree, a new node is created with both of its pointers pointing to the empty node. All newly created nodes start with the frozen flag set to false (temporarily mutable). At this point, the tree is made immutable by setting the frozen flag to true as indicated on the right side of the figure. This makes the tree immutable for the rest of its lifetime.

- While adding other items, the only way to add the second item is to not only create a node for it but also create another node for the first item because the previous node is immutable, you can’t change its pointers.
